### ISSUE TITLE
Faciliate monitoring and guiding of OSPRs

### ISSUE THREAD

**Issue by antoviaque** [openedx/wg-coordination#103]
See https://discuss.openedx.org/t/pull-requests-review-delays/10497 for the context

- Monitoring PRs is time-consuming: Making sure every two weeks that maintainers have responded to specific ones might be a bit of a challenge to keep track of
- Facilitate quickly identify which PRs would qualify for review by any developer - ideally this wouldn’t require any manual combing, a filter can display them on a board
	* OSPRs that were opened two months ago or earlier: https://github.com/orgs/openedx/projects/19/views/22?filterQuery=date-opened%3A%3C%3D2023-06-16+-status%3ADone
	* OSPRs that became ready for review a month ago or earlier: https://github.com/orgs/openedx/projects/19/views/22?filterQuery=ready-for-review-since%3A%3C%3D2023-07-16+-status%3ADone

- Link repos to specific people:
  - For some repos, like edx-platform, it would be good to nail down exactly which CCs have rights to review/merge in what specific areas
  - 2U has a basic breakdown of which 2U teams own certain projects on edx-platform, but it’s not as specific as we need it to be

- Provide better information about the current situation on the different repos, since that can be different
	- => to link to the effort to document who the maintainers are?
	- Ned: We have a CODEOWERS, but it’s incomplete, and GitHub points out that it has errors
		-  Xavier: Do you know if the file currently maintained? And how does that relate (if it does) with the process to identify maintainers in 0001 Use Backstage to Support Maintainers — Open edX Proposals 1.0 documentation ?
	- review bot could use a list of core contributors for the code
	-  Jenna is also building a list of relationships between features ↔ repositories

**Comment by itsjeyd** [comment]
Closely related: #104

**Comment by e0d** [comment]
Oh nice, that's a very useful query and useful views, thanks for doing this.

**Comment by itsjeyd** [comment]
@mphilbrick211 Hm... I might be missing something, but with the scenario that you're describing I don't think PRs would be shown in the "stuck for 14 days" views: If a PR gets reviewed (= commented on), that should bump the `last updated` field to a date less than 14 days ago. So it *should* disappear from the "stuck for 14 days" views. Are there any specific examples where that didn't happen?

**Comment by mphilbrick211** [comment]
> @mphilbrick211 Hm... I might be missing something, but with the scenario that you're describing I don't think PRs would be shown in the "stuck for 14 days" views: If a PR gets reviewed (= commented on), that should bump the `last updated` field to a date less than 14 days ago. So it _should_ disappear from the "stuck for 14 days" views. Are there any specific examples where that didn't happen?

This is a good point - I didn't remember we had that field that would auto-update.

**Comment by itsjeyd** [comment]
🚧 WIP 🚧 

### Challenges

> * Link repos to specific people:
>     * For some repos, like edx-platform, it would be good to nail down exactly which CCs have rights to review/merge in what specific areas
>     * 2U has a basic breakdown of which 2U teams own certain projects on edx-platform, but it’s not as specific as we need it to be
>
> * Provide better information about the current situation on the different repos, since that can be different
>     * => to link to the effort to document who the maintainers are?
>     * Ned: We have a [CODEOWNERS](https://github.com/openedx/edx-platform/blob/master/.github/CODEOWNERS), but it’s incomplete, and GitHub points out that it has errors
>         * Xavier: Do you know if the file currently maintained? And how does that relate (if it does) with the process to identify maintainers in [0001 Use Backstage to Support Maintainers](https://open-edx-proposals.readthedocs.io/en/latest/processes/oep-0055/decisions/0001-use-backstage-to-support-maintainers.html) — Open edX Proposals 1.0 documentation ?
>     * review bot could use a list of core contributors for the code
>     * Jenna is also building a list of relationships between features ↔ repositories

### Comments and additional details

[Ref](https://discuss.openedx.org/t/contributors-meetup-async-update-september-2nd-2023-september-15th-2023/11185/6):

> ... currently it is a bit tricky to tell who to ping on some parts of the code. Maybe having to specify this to a level that a bot is able to ping people on its own would help simplify things, too?

[Ref](https://discuss.openedx.org/t/contributors-meetup-async-update-september-2nd-2023-september-15th-2023/11185/8):

> Right now, information about who to ping is spread out over several resources -- a few public wiki pages about the [CC](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3156344833/Core+Contributors+to+the+Open+edX+Project) and [maintainership](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3663233055/Cumulative+Maintainer+Responsibilities) programs, [Open edX Backstage](https://backstage.openedx.org/), and a spreadsheet that is available to Michelle and me but not the wider community. For maintainership specifically, information from the wiki and Open edX Backstage overlaps to a large degree but last I checked the wiki seemed to be a bit ahead of Backstage.

> If we could unify the different resources somehow and find a way to make the information that they provide more discoverable (e.g. by having the bot surface it on PRs like you're suggesting), that would be great -- it would not only help Michelle and me in the context of OSPR management, but also empower PR authors to get in touch with reviewers themselves.

[Ref](https://discuss.openedx.org/t/contributors-meetup-async-update-september-2nd-2023-september-15th-2023/11185/12):

> If anything, I think we need to make sure you both have less work to do, by automating things like pinging the right reviewers. Imho it would be a better use of your time to allow you to focus on things like escalating PRs that don’t get attention, or monitoring the time it takes to merge PRs to proactively suggest improvements – ie things that actually require human attention and inventiveness.



Essentially, we want to:

- refine information about code ownership
- improve discoverability of information about code ownership

We also want to continue assigning code ownership (via the maintainership program). Fortunately, that's already happening and we can consider it out of scope here.

Ultimately, we would like to enable the https://github.com/openedx-webhooks GitHub bot to utilize information about code ownership so that we can automate certain steps of the OSPR management process ([ref](https://discuss.openedx.org/t/contributors-meetup-async-update-september-2nd-2023-september-15th-2023/11185/6)):

> ... for example having the bot suggesting/assigning/pinging reviewers based on the code that’s being modified

So we need to:

> ... find a way to unify these resources, and write a proposed specification for the bot to automate [surfacing relevant information on OSPRs]

However, we will need to [take into account](https://discuss.openedx.org/t/contributors-meetup-async-update-september-2nd-2023-september-15th-2023/11185/14) that:

> [Michelle and Tim] have a high-level understanding of the available resources and how to use them to get the info that [they] need for OSPR management. But [we] don’t have a detailed list of things that are currently missing, redundant, or maybe even inconsistent. So as a next step it would make sense to have a closer look and document that.

### Review of existing resources

#### Open edX wiki

##### CC Program

**Main resource**: [Core Contributors to the Open edX Project](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3156344833/Core+Contributors+to+the+Open+edX+Project)

- Comprehensive list of current (and past) Core Contributors.
- Documents write access to repos for CCs that are Code Contributors, but format isn't standardized.
- ...

##### Maintainership Pilot

**Main resource**: [Cumulative Maintainer Responsibilities](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3663233055/Cumulative+Maintainer+Responsibilities)

...

#### Open edX Backstage

https://backstage.openedx.org/

...

#### Other resources

* 2U spreadsheet with code ownership mapping (private, shared with OSPR managers)

...

### Potential solutions

TBD

### Changes applied

TBD

**Comment by antoviaque** [comment]
@itsjeyd How are things going with this issue? Anything blocking I can help with?

Note that I don't know if we currently track product review delays as part of the current ticket, but delays there too have been mentioned by Nacho in the last Campus meeting: https://otter.ai/u/IkmOT8IcZ7XWmO3x_DZBVeg1HyU?tab=comments&view=transcript&t=1487s -- if it's not in the scope of the current ticket, it would be worth creating a ticket to track it.

**Comment by itsjeyd** [comment]
@antoviaque Thanks for checking in! There are no blockers per se, I simply haven't found time to continue working on this issue yet.

And you're right, product review delays aren't in scope here, and creating a separate ticket for addressing them would make sense. It [sounds like](https://otter.ai/s/GeWrFOTyQX2OeURJxV27Ig?snpt=true) there is consensus that the process for product reviews is due for an update. So if we could find someone with the necessary bandwidth to drive that conversation forward in the coming weeks/months, it should be possible to make some improvements.

CC @mphilbrick211

**Comment by mphilbrick211** [comment]
Hi @antoviaque and @itsjeyd - Product has started a new process (very recently) where they are trying to catch and review product-related contributions **_before_** they go onto the Contributions board at all. At the moment, [here are the current items they're reviewing](https://openedx.atlassian.net/wiki/spaces/OEPM/pages/3901390856/Open+Source+Contributions+-+Product+Proposals).

Once product is done reviewing an item, they will move it to the Contributions board with a label saying "Product Review Complete" (we haven't created that label yet as this is still brand new). This way, once it's on the board, we know it's waiting for Engineering only. Given that this is something we're trying out, it might be a bit rocky for the first few months until we get a good rhythm going.

There are still a few stalled product items from over the past several months, but some of them require lots of extra discovery work, so those PRs have been moved to drafts until it's more appropriate to move forward with them (so, intentionally paused, not stalled).

@itsjeyd - can you let me know which Product items are stalled on your side?
@antoviaque - are there any you're worried about that are stalled that maybe I can help get you an update on? I'm with Jenna in the office this week, so I can ask.

**Comment by itsjeyd** [comment]
@mphilbrick211 

> can you let me know which Product items are stalled on your side?

Will do. I have my next round of triage coming up on Thursday.

**Comment by antoviaque** [comment]
@mphilbrick211 Thanks for the update on this! And for the product reviews that are stalled, Jenna would already be aware since Nacho brought that to her in the Campus meeting, so I think that's good, but I'll let you know if I hear about another, besides what @itsjeyd would already know about.

**Comment by jalondonot** [comment]
20/02 Contributor's meetup update:
- The survey results will be reviewed this week

**Comment by antoviaque** [comment]
Referencing the discussion from https://openedx.atlassian.net/wiki/spaces/COMM/pages/4128014347/2024-03-21+Meeting+notes?focusedCommentId=4130308114 which is related to this ticket : "Should 2U PRs go to the Contributions Board?"

**Comment by antoviaque** [comment]
@itsjeyd To follow-up on the discussion from https://github.com/openedx/wg-coordination/issues/104#issuecomment-2220284523 on the ticket that currently tracks this type of improvements:

> The issue that I mentioned [above](https://github.com/openedx/wg-coordination/issues/104#issuecomment-2128973336) (= no mapping between repos and CCs) is still open, though. I had posted about it [internally](https://tasks.opencraft.com/browse/MNG-2649?focusedCommentId=292256&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-292256) before the conference but there were no takers so far.

Did you end up getting the help you need?

**Comment by itsjeyd** [comment]
@antoviaque Yep, we have a [discovery ticket](https://tasks.opencraft.com/browse/BB-9084) that @xitij2000 is working on. He has already come up with a potential solution and [posted about it](https://openedx.slack.com/archives/C03R320AFJP/p1725025364346349) in the #wg-maintenance channel on Slack to get feedback on it. CC @mphilbrick211  

Thanks for checking in! :)

**Comment by antoviaque** [comment]
@itsjeyd Great, thanks! :+1: 

@xitij2000 I see that you didn't get many answers on Slack - it could be worth posting it in the forum too - I don't think everyone sees it in Slack, which is more synchronous. You should also ping specific people that you want feedback from on this - and maybe attend a maintainers working group meeting to ask for reviewers. Shout if you need help.

**Comment by xitij2000** [comment]
> @itsjeyd Great, thanks! 👍
> 
> @xitij2000 I see that you didn't get many answers on Slack - it could be worth posting it in the forum too - I don't think everyone sees it in Slack, which is more synchronous. You should also ping specific people that you want feedback from on this - and maybe attend a maintainers working group meeting to ask for reviewers. Shout if you need help.

The meetings are at a difficult time for me, but I was planning on attending one since I didn't get any responses. 

I'll post on the forum first and see if that fares better.

**Comment by itsjeyd** [comment]
@antoviaque @xitij2000 I'll be attending maintenance working group meetings again from next week on, so should be getting an opportunity soon to raise the topic there.

The more places we bring this up, the better.

**Comment by xitij2000** [comment]
@itsjeyd Thanks! I think the general impression I get from the forum discussions is that we should automate adding this to catalog-info.yaml and then from there it can be pulled into Backstage. This will allow anyone to open this file from the repo and see the list of contributors. 

What I liked about the project board is that it allows one to quickly filter based on either the repo or the contributor, however, I can understand how it might not be ideal either since project boards aren't really designed for this purpose.

**Comment by itsjeyd** [comment]
@xitij2000 I just [replied](https://discuss.openedx.org/t/improving-discoverability-of-core-contributors/13862/17) to the conversation over on the forum thread -- maybe there's a way to integrate CC info into the [Contributions](https://github.com/orgs/openedx/projects/19) board? That would be a bit of a game changer for OSPR management.

**Comment by antoviaque** [comment]
There are a number of other tickets that aim to move this topic forward:
* https://github.com/openedx/wg-coordination/issues/134 (and its many subtasks, see ticket)
* https://github.com/openedx/wg-coordination/issues/135
* https://github.com/openedx/wg-coordination/issues/138

I'm going to move the current task to the "blocked" column, to focus updates on the related tasks this depends on - we can come back to this task once they are completed, to figure out if the overall goal was attained.
